File: problem14/p14.py
def chain_length(number, limit, chain):
    length = 1
    copyNum = number
    while number != 1:
        if number < limit:
            if chain[number] > 0:
                length += chain[number]
                break
        if number % 2 == 0:
            number = number // 2
        else:
            number = (number * 3) + 1
        length += 1
    chain[copyNum] = length
    return length


def find_longest_chain(limit):
    chain = [0] * limit
    max = 0
    for num in range(1, limit):
        length = chain_length(num, limit, chain)
        if length > max:
            max = length
            perfectNum = num
    return perfectNum


print(find_longest_chain(1000000))

File: problem14/p14.java
public class p14 {
	
	public static int chain_length(long number, int limit, int[] chain) {
	    int length = 1;
	    long copyNumber = number;
	    while(number != 1){
	        if(number < limit){
	            if(chain[(int) number] != 0){
	            length += chain[(int) number];
	            break;
	            }
	        }
	        if(number % 2 == 0)
	            number /= 2;
	        else
	            number = (number * 3) + 1;
	        length++;
	    }
	    chain[(int) copyNumber] = length;
	    return length;
	}
	
	public static int find_longest_chain(int limit) {
	    int[] chain = new int[limit];
	    int i, max = 0, length, initialNum = 0;
	    for(i=1; i<limit; i++){
	        length = chain_length(i, limit, chain);
	        if(length > max){
	            max = length;
	            initialNum = i;
	        }
	    }

	    return initialNum;
	}

	public static void main(String[] args) {
		
		System.out.println(find_longest_chain(1000000));
		
	}
	
}
